File: packages/sdk/src/agent.ts
import bs58 from "bs58";
import nacl from "tweetnacl";

export type TaskType = "fast" | "reasoning" | "general";

export type Receipt = {
  agent_pubkey: string;
  provider: string;
  model: string;
  cost_usdc: number;
  balance_remaining: number;
  prompt_hash: string;
  response_hash: string;
  timestamp: number;
  inference_id: string | null;
  points_total: number;
};

export type InferenceResponse = {
  ok: boolean;
  result?: string;
  receipt?: Receipt;
  error?: string;
  detail?: string;
};

export type InferenceOptions = {
  prompt: string;
  task_type?: TaskType;
  max_cost_usdc?: number;
  /**
   * When the prepaid balance is empty, automatically request a sponsored
   * grant from `${endpoint}/grants` and retry once. Default `true` — this
   * is the zero-friction first-call path for new agents. Set to `false`
   * to fail fast on insufficient credits without the extra grant probe.
   */
  autoGrant?: boolean;
};

export type GrantResponse = {
  ok: boolean;
  agent_pubkey?: string;
  grant_usdc?: number;
  balance_usdc?: number;
  error?: string;
  detail?: string;
};

const DEFAULT_TASK_TYPE: TaskType = "general";

export class Agent {
  readonly publicKey: Uint8Array;
  readonly secretKey: Uint8Array;

  constructor(secretKey: Uint8Array) {
    if (secretKey.length !== 64) {
      throw new Error("secretKey must be 64 bytes (Ed25519 seed + public key)");
    }
    this.secretKey = secretKey;
    this.publicKey = secretKey.slice(32);
  }

  static generate(): Agent {
    const kp = nacl.sign.keyPair();
    return new Agent(kp.secretKey);
  }

  static fromBase58(secretKeyBase58: string): Agent {
    return new Agent(bs58.decode(secretKeyBase58));
  }

  get pubkey(): string {
    return bs58.encode(this.publicKey);
  }

  get secretKeyBase58(): string {
    return bs58.encode(this.secretKey);
  }

  signBody(body: string): string {
    const bytes = new TextEncoder().encode(body);
    const sig = nacl.sign.detached(bytes, this.secretKey);
    return bs58.encode(sig);
  }

  async inference(
    endpoint: string,
    opts: InferenceOptions
  ): Promise<InferenceResponse> {
    const base = endpoint.replace(/\/$/, "");
    const autoGrant = opts.autoGrant !== false; // default on

    const first = await this._inferenceOnce(base, opts);
    if (first.ok) return first;

    // The grant endpoint exists to remove the "buy USDC before the first
    // call" friction. If the agent has zero balance, probe it once. A
    // grant failure (already issued, IP cap, budget exhausted) is silent —
    // the user gets the original insufficient_credits error back, which is
    // the same outcome they would have had without auto-grant.
    if (autoGrant && first.error === "insufficient_credits") {
      const grant = await this.grant(base);
      if (grant.ok) {
        return await this._inferenceOnce(base, opts);
      }
    }
    return first;
  }

  /**
   * Request a sponsored USDC grant for this agent's pubkey from
   * `${endpoint}/grants`. One grant per pubkey ever, capped per-IP and per
   * global daily budget by the operator. Safe to call explicitly even if
   * `inference()` already auto-grants — the server returns 409
   * `grant_already_issued` with the current balance.
   */
  async grant(endpoint: string): Promise<GrantResponse> {
    const base = endpoint.replace(/\/$/, "");
    try {
      const res = await fetch(`${base}/grants`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ agent_pubkey: this.pubkey }),
      });
      return await parseJsonResponse<GrantResponse>(res, `${base}/grants`);
    } catch (e) {
      return {
        ok: false,
        error: "grant_request_failed",
        detail: e instanceof Error ? e.message : "unknown",
      };
    }
  }

  private async _inferenceOnce(
    base: string,
    opts: InferenceOptions
  ): Promise<InferenceResponse> {
    const body = JSON.stringify({
      prompt: opts.prompt,
      task_type: opts.task_type ?? DEFAULT_TASK_TYPE,
      nonce: cryptoRandomUUID(),
      timestamp: Date.now(),
      ...(opts.max_cost_usdc !== undefined
        ? { max_cost_usdc: opts.max_cost_usdc }
        : {}),
    });
    const signature = this.signBody(body);

    const res = await fetch(`${base}/inference`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Agent-Pubkey": this.pubkey,
        "X-Nexus-Signature": signature,
      },
      body,
    });

    return await parseJsonResponse<InferenceResponse>(res, `${base}/inference`);
  }
}

/**
 * Parse a fetch Response that is expected to carry a JSON body.
 *
 * Endpoints normally return JSON on every path, but infrastructure in
 * front of them (load balancer, serverless platform, crashed handler)
 * can produce an empty or non-JSON body — historically that surfaced as
 * a bare `SyntaxError: Unexpected end of JSON input` with no HTTP
 * context. Instead, return a structured `{ ok: false }` object that
 * names the status code, the endpoint, and a snippet of whatever body
 * came back, so callers can log and display something actionable.
 */
async function parseJsonResponse<T extends { ok: boolean; error?: string; detail?: string }>(
  res: Response,
  endpoint: string
): Promise<T> {
  const text = await res.text();
  if (text.length > 0) {
    try {
      return JSON.parse(text) as T;
    } catch {
      return {
        ok: false,
        error: "upstream_non_json",
        detail: `HTTP ${res.status} from ${endpoint} returned a non-JSON body: ${text.slice(0, 200)}`,
      } as T;
    }
  }
  return {
    ok: false,
    error: "upstream_empty_body",
    detail: `HTTP ${res.status} from ${endpoint} returned an empty body`,
  } as T;
}

function cryptoRandomUUID(): string {
  const g = globalThis as { crypto?: { randomUUID?: () => string } };
  if (g.crypto?.randomUUID) return g.crypto.randomUUID();
  const bytes = nacl.randomBytes(16);
  bytes[6] = (bytes[6]! & 0x0f) | 0x40;
  bytes[8] = (bytes[8]! & 0x3f) | 0x80;
  const hex = Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

File: packages/sdk-python/src/vdm_nexus/agent.py
"""Python port of @vdm-nexus/sdk's Agent class."""

from __future__ import annotations

import json
import time
import uuid
from typing import Any, Literal, Optional, TypedDict

import base58
import httpx
from nacl.signing import SigningKey

from .sign import sign_body as _sign

TaskType = Literal["fast", "reasoning", "general"]
_DEFAULT_TASK_TYPE: TaskType = "general"


class Payment(TypedDict, total=False):
    scheme: str
    network: str
    pay_to: str
    amount_usdc: float
    tx_signature: str


class Receipt(TypedDict, total=False):
    # SIR v2 fields — `v`, `nexus_signature`, and (for x402 receipts)
    # `payment` were missing before 2026-05-23. They've always been on
    # the wire; the type just didn't reflect them. See the SIR v2 spec
    # at docs.vdmnexus.com/docs/spec/sir-v2 for the canonical shape.
    v: int
    agent_pubkey: str
    provider: str
    upstream: str
    model: str
    cost_usdc: float
    balance_remaining: float
    prompt_hash: str
    response_hash: str
    timestamp: int
    inference_id: Optional[str]
    points_total: int
    payment: Payment
    nexus_signature: str


class InferenceResponse(TypedDict, total=False):
    ok: bool
    result: str
    receipt: Receipt
    error: str
    detail: str


class GrantResponse(TypedDict, total=False):
    ok: bool
    agent_pubkey: str
    grant_usdc: float
    balance_usdc: float
    error: str
    detail: str


class Agent:
    """Autonomous agent identified by an Ed25519 keypair.

    Mirrors `packages/sdk/src/agent.ts`. The Solana-compatible 64-byte
    secret key (seed || public key) is the sole credential — no API keys.
    """

    def __init__(self, secret_key: bytes) -> None:
        if len(secret_key) != 64:
            raise ValueError("secret_key must be 64 bytes (Ed25519 seed + public key)")
        self.secret_key = bytes(secret_key)
        self.public_key = self.secret_key[32:]

    @classmethod
    def generate(cls) -> "Agent":
        sk = SigningKey.generate()
        return cls(bytes(sk) + bytes(sk.verify_key))

    @classmethod
    def from_base58(cls, secret_key_base58: str) -> "Agent":
        return cls(base58.b58decode(secret_key_base58))

    @property
    def pubkey(self) -> str:
        """Base58-encoded public key. This is the agent's identity."""
        return base58.b58encode(self.public_key).decode("ascii")

    @property
    def secret_key_base58(self) -> str:
        """Base58-encoded 64-byte secret key. Treat as a password."""
        return base58.b58encode(self.secret_key).decode("ascii")

    def sign_body(self, body: bytes) -> str:
        """Sign the raw body bytes; returns a base58 signature.

        Pass the EXACT bytes you will send as the HTTP body — the Nexus
        server re-derives the signature against the bytes it receives.
        """
        return _sign(self.secret_key, body)

    async def inference(
        self,
        endpoint: str,
        *,
        prompt: str,
        task_type: TaskType = _DEFAULT_TASK_TYPE,
        max_cost_usdc: Optional[float] = None,
        auto_grant: bool = True,
        client: Optional[httpx.AsyncClient] = None,
    ) -> InferenceResponse:
        """POST a signed inference request to `<endpoint>/inference`.

        On `insufficient_credits`, optionally probes `<endpoint>/grants`
        once and retries. Pass `auto_grant=False` to fail fast instead.
        Pass `client` to reuse an `httpx.AsyncClient` for pooling.
        """
        base = endpoint.rstrip("/")
        opts = {"prompt": prompt, "task_type": task_type, "max_cost_usdc": max_cost_usdc}

        async def go(c: httpx.AsyncClient) -> InferenceResponse:
            first = await self._inference_once(c, base, opts)
            if first.get("ok"):
                return first
            if auto_grant and first.get("error") == "insufficient_credits":
                grant = await self.grant(base, client=c)
                if grant.get("ok"):
                    return await self._inference_once(c, base, opts)
            return first

        if client is not None:
            return await go(client)
        async with httpx.AsyncClient() as owned:
            return await go(owned)

    async def grant(
        self,
        endpoint: str,
        *,
        client: Optional[httpx.AsyncClient] = None,
    ) -> GrantResponse:
        """Request a sponsored USDC grant for this agent's pubkey."""
        url = f"{endpoint.rstrip('/')}/grants"
        body = json.dumps({"agent_pubkey": self.pubkey}, separators=(",", ":")).encode()
        headers = {"Content-Type": "application/json"}
        try:
            if client is not None:
                res = await client.post(url, content=body, headers=headers)
            else:
                async with httpx.AsyncClient() as c:
                    res = await c.post(url, content=body, headers=headers)
            return _parse_json(res)  # type: ignore[return-value]
        except (httpx.HTTPError, ValueError) as e:
            return {"ok": False, "error": "grant_request_failed", "detail": str(e)}

    async def _inference_once(
        self,
        client: httpx.AsyncClient,
        base: str,
        opts: dict[str, Any],
    ) -> InferenceResponse:
        payload: dict[str, Any] = {
            "prompt": opts["prompt"],
            "task_type": opts.get("task_type") or _DEFAULT_TASK_TYPE,
            "nonce": str(uuid.uuid4()),
            "timestamp": int(time.time() * 1000),
        }
        if opts.get("max_cost_usdc") is not None:
            payload["max_cost_usdc"] = opts["max_cost_usdc"]
        # CRITICAL: serialize once. The same bytes get signed AND sent.
        body = json.dumps(payload, separators=(",", ":")).encode("utf-8")
        res = await client.post(
            f"{base}/inference",
            content=body,
            headers={
                "Content-Type": "application/json",
                "X-Agent-Pubkey": self.pubkey,
                "X-Nexus-Signature": self.sign_body(body),
            },
        )
        return _parse_json(res)  # type: ignore[return-value]


def _parse_json(res: httpx.Response) -> dict[str, Any]:
    try:
        return res.json()
    except ValueError:
        return {
            "ok": False,
            "error": "invalid_response",
            "detail": f"non-JSON response (status {res.status_code})",
        }
